woocommerce/woocommerce-ios
# [Bundle products in order form] Adjust bundle item quantity based on the parent item quantity

Closes: #11136


## Why


When the bundle product order item's quantity is increased to n (n > 1) **and then configuring the bundle** (see the detailed behavior in Pe5pgL-3Vd-p2#quantity-of-bundle-child-order-items), the child bundled items have their quantity multiplied by the parent bundle quantity. Before this PR, we set the bundled item's quantity in the Configure form to the quantity in the order item. However, this multiplied quantity doesn't match the actual quantity **per bundle** and should be divided by the parent order item's quantity. The multiplied quantity of the bundled item can also exceed the maximum quantity and cause confusion.

## How

To match the behavior in the extension (link in Pe5pgL-3Vd-p2#quantity-of-bundle-child-order-items), the parent bundle order item is now DI'ed to `ConfigurableBundleProductViewModel` -> `ConfigurableBundleItemViewModel` so that the bundled item's quantity is divided by the parent order item's quantity (if smaller than 1, it's set to 1 to avoid dividing by zero). Test cases were added to verify this behavior.

## Testing instructions


Prerequisite: the store has [Product Bundles extension](https://woocommerce.com/products/product-bundles/) enabled, and also a bundle product with non-empty items. **The store has an order with a bundle product of quantity 1 with non-empty items.**

🗒️ Please note that there's a known issue where adding a bundle from the product selector without making any changes doesn't work. If you need to create an order with a bundle, please do it in core or make some changes to the bundle before tapping `Done`.

- Go to the Orders tab
- Tap on the order in the prerequisite
- Tap `Edit`
- Increment the parent bundle order item's quantity to like 3
- Tap to expand the parent bundle order item, then tap `Configure` --> the bundle items should have the same quantity as before
- Update the quantity of a bundle item, then tap `Done` --> in the order form, the bundled child order items should have the latest quantity multiplied by the parent quantity after the remote update
- Tap to expand the parent bundle order item, then tap `Configure` --> the bundle items should have the same quantity as in the last configuration (before this PR, the multiplied quantity is shown)

## Screenshots


order details | configure form
-- | --
<img src="https://github.com/woocommerce/woocommerce-ios/assets/1945542/6d5f4d07-e234-4aa2-94e4-c4a3bfe560cd" width="300" /> | <img src="https://github.com/woocommerce/woocommerce-ios/assets/1945542/db063545-b9e1-4ebd-b79d-4ee4afbe9dbf" width="300" />

---
